File: Dataset/Leetcode/train/121/182.cpp
 class Solution {
public:
    int XXX(vector<int>& prices) {
        int n = prices.size();
        //数组下标表示在第i天前所能获得的最大利润
        vector<int>dp(n);
        int minprice = prices[0];//记录当前下标之前的最小进价
        for(int i = 1;i<n;++i)
        {
            //递推公式 
            //第i天获得迪润最大值为（前一天利润最大值和今天卖出获得利润）的最大值
            //即dp[i] = max(dp[i-1],prices[i]-minprice);
            dp[i] = max(dp[i-1],prices[i]-minprice);
            minprice = min(minprice,prices[i]);
        }
        return dp[n-1];
    }
};

File: Dataset/Leetcode/train/110/182.java
class Solution {
    public boolean XXX(TreeNode root) {
        if(root == null){
            return true;
        }

        int gap = height(root.left) - height(root.right);

        return -1<=gap && gap<=1 && XXX(root.left) && XXX(root.right);

    }

    public int height(TreeNode root){
        if(root == null){
            return 0;
        }

        return Math.max(height(root.left), height(root.right)) + 1;
    }
}

